File: light-control/algorithms/_meta/base_algorithm.py
import logging
from .pixel import Pixel
from .helpers import get_millis

class BaseAlgorithm:
    def __init__(self, num_pixels, alg_config, settings) -> None:
        self.logger = logging.getLogger(__name__)
        self.name = alg_config['name']
        self.logger.log(logging.INFO, f'Initializing algorithm {self.name:s}')
        self.config = alg_config
        self.__settings = settings
        self.__last_cycle_time = 0

        self.pixels = [Pixel() for _ in range(num_pixels)]
    
    def refresh_rate(self):
        return self.__settings.get('refresh_rate', self.config['refresh_rate'])

    def settings(self):
        return self.__settings

    def run_cycle(self, elapsed_millis:float, elapsed_seconds:float) -> bool:
        pass

File: light-control-ts/src/algorithms/_meta/baseAlgorithm.ts
import { Pixel } from './pixel.js'

export interface AlgorithmConfig {
  name: string
  refresh_rate: number
  options: Record<string, unknown>
}

export class BaseAlgorithm {
  readonly name: string
  pixels: Pixel[]
  protected readonly config: AlgorithmConfig
  private readonly algSettings: Record<string, unknown>

  constructor(numPixels: number, config: AlgorithmConfig, settings: Record<string, unknown>) {
    this.name = config.name
    this.config = config
    this.algSettings = settings
    this.pixels = Array.from({ length: numPixels }, () => new Pixel())
  }

  refreshRate(): number {
    return (this.algSettings['refresh_rate'] as number | undefined) ?? this.config.refresh_rate
  }

  settings(): Record<string, unknown> {
    return this.algSettings
  }

  // Returns true when the algorithm signals completion (used by transition)
  runCycle(_elapsedMillis: number, _elapsedSeconds: number): boolean {
    return false
  }
}
